### ISSUE TITLE
Stackoverflow in MultiResponseBuilder

### ISSUE THREAD

**Issue by DivineTraube** [xetorthio/jedis#856]
Hi,

if one performs many operations (a few thousand) in a multi over a pipeline, the recursive nature of the response processing in Jedis will always cause a Stackoverflow.

Code to reproduce this in Jedis 2.6.2 :
```java
Jedis jedis = new Jedis("localhost");
Pipeline p = jedis.pipelined();
p.multi();
for (int i = 0; i < 10000; i++) {
            p.setbit("test", 1, true); //or any other call
 }
Response<List<Object>> exec = p.exec();
p.sync();
exec.get();  //Booom
```

Stacktrace:
```
java.lang.StackOverflowError
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:35)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:31)
	at redis.clients.jedis.Pipeline$MultiResponseBuilder.build(Pipeline.java:12)
	at redis.clients.jedis.Response.build(Response.java:50)
	at redis.clients.jedis.Response.get(Response.java:28)
```

**Comment by marcosnils** [comment]
@DivineTraube as you may already know this is a common issue when dealing with recursive calls in java as the thread stack size has a small default value (1024k).

The example above makes 50k recursive calls without allocating any variables inside the stack and also throws a a StackOverFlow error.

```java
public class MainTest {
    
    public static int counter = 0;

    public static void main(String[] args) {
	sum();
    }
    
    private static void sum() {
	if (counter++ <= 50000) {
	    sum();
	}
    }
}
```

If you're planning to make a lot of pipeline calls I'd recommend to increase the thread stack size using the -Xss jvm parameter to suit your needs. Per your example a value of -Xss2048k should be enough to make 10k pipeline calls.

**Comment by HeartSaVioR** [comment]
@DivineTraube 
Thanks for reporting. I'll investigate on it.
@marcosnils 
Maybe we have to find a way to take care of dependency without cross reference, which makes recursion. Please share any ideas or approaches anytime when you'll have.

**Comment by HeartSaVioR** [comment]
Maybe I got fixed it by #860.

I can't remove cross reference cause we cannot know what response.get() is called first.
(It can be "exec"'s response, or "request in multi"'s response.)
Instead, I changed Response.build() to prevent recursion by checking build is in progress.

**Comment by DivineTraube** [comment]
Great, merging this would be highly appreciated

**Comment by marcosnils** [comment]
Merged to master, 2.7 and 2.6
